src/sources/mithril.rs

```diff
@@ -0,0 +1,249 @@
+            }
+        }
+    }
+}
+
+async fn fetch_snapshot(config: &Config, feedback: Arc<Feedback>) -> MithrilResult<()> {
+    let client = ClientBuilder::aggregator(&config.aggregator, &config.genesis_key)
+        .add_feedback_receiver(feedback)
+        .build()?;
+
+    let snapshots = client.snapshot().list().await?;
+
+    let last_digest = snapshots
+        .first()
+        .ok_or(MithrilError::msg("no snapshot available"))?
+        .digest
+        .as_ref();
+
+    let snapshot = client
+        .snapshot()
+        .get(last_digest)
+        .await?
+        .ok_or(MithrilError::msg("no snapshot available"))?;
+
+    let target_directory = Path::new(&config.snapshot_download_dir);
+
+    client
+        .snapshot()
+        .download_unpack(&snapshot, target_directory)
+        .await?;
+
+    if let Err(e) = client.snapshot().add_statistics(&snapshot).await {
+        warn!("failed incrementing snapshot download statistics: {:?}", e);
+    }
+
+    let certificate = if config.skip_validation {
+        client
+            .certificate()
+            .get(&snapshot.certificate_hash)
+            .await?
+            .ok_or(MithrilError::msg("certificate for snapshot not found"))?
+    } else {
+        client
+            .certificate()
+            .verify_chain(&snapshot.certificate_hash)
+            .await?
+    };
+
+    let message = MessageBuilder::new()
+        .compute_snapshot_message(&certificate, target_directory)
+        .await?;
+
+    assert!(certificate.match_message(&message));
+
+    Ok(())
+}
+
+#[derive(Stage)]
+#[stage(name = "source", unit = "()", worker = "Worker")]
+pub struct Stage {
+    config: Config,
+
+    pub output: SourceOutputPort,
+}
+
+pub struct Worker {
+    config: Config,
+}
+
+impl Worker {}
+
+#[async_trait::async_trait(?Send)]
+impl gasket::framework::Worker<Stage> for Worker {
+    async fn bootstrap(stage: &Stage) -> Result<Self, WorkerError> {
+        let feedback = Arc::new(Feedback::default());
+        let target_directory = Path::new(&stage.config.snapshot_download_dir);
+
+        if !target_directory.exists() {
+            std::fs::create_dir_all(target_directory)
+                .map_err(|err| miette::miette!(err.to_string()))
+                .context(format!(
+                    "Failed to create directory: {}",
+                    target_directory.display()
+                ))
+                .map_err(|_| WorkerError::Panic)?;
+        }
+
+        fetch_snapshot(&stage.config, feedback.clone())
+            .await
+            .map_err(|err| miette::miette!(err.to_string()))
+            .context("fetching and validating mithril snapshot")
+            .map_err(|_| WorkerError::Panic)?;
+
+        Ok(Self {
+            config: stage.config.clone(),
+        })
+    }
+
+    async fn schedule(&mut self, _stage: &mut Stage) -> Result<WorkSchedule<()>, WorkerError> {
+        Ok(WorkSchedule::Unit(()))
```

Comment:
This will reimport the snapshot in a loop. You need to send a WorkSchedule::Done after you're done processing the snapshot.

Reply:
thanks for the feedback, makes sense!